### ISSUE TITLE
[BUG] spring-cloud-azure-starter-active-directory:4.4.0 OAuth 2.0 flow failing with UnsupportedOperationException in JacksonHttpSessionOAuth2AuthorizedClientRepository

### ISSUE THREAD

**Issue by mtter** [Azure/azure-sdk-for-java#31256]
**Describe the bug**

We are using  `com.azure.spring:spring-cloud-azure-starter-active-directory` to authenticate users using OAuth 2.0 in a Spring Boot application. After upgrading `spring-cloud-azure-starter-active-directory` from version 4.3.0 to version 4.4.0 (no other changes) the login flow fails with an `UnsupportedOperationException` in `JacksonHttpSessionOAuth2AuthorizedClientRepository`.

***Exception or Stack Trace***
```
java.lang.UnsupportedOperationException: null
	at java.base/java.util.AbstractMap.put(AbstractMap.java:209)
	at com.azure.spring.cloud.autoconfigure.aad.implementation.oauth2.JacksonHttpSessionOAuth2AuthorizedClientRepository.saveAuthorizedClient(JacksonHttpSessionOAuth2AuthorizedClientRepository.java:52)
	at org.springframework.security.oauth2.client.web.OAuth2LoginAuthenticationFilter.attemptAuthentication(OAuth2LoginAuthenticationFilter.java:204)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:227)
	at org.springframework.security.web.authentication.AbstractAuthenticationProcessingFilter.doFilter(AbstractAuthenticationProcessingFilter.java:217)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.oauth2.client.web.OAuth2AuthorizationRequestRedirectFilter.doFilterInternal(OAuth2AuthorizationRequestRedirectFilter.java:178)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:103)
	at org.springframework.security.web.authentication.logout.LogoutFilter.doFilter(LogoutFilter.java:89)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.header.HeaderWriterFilter.doHeadersAfter(HeaderWriterFilter.java:90)
	at org.springframework.security.web.header.HeaderWriterFilter.doFilterInternal(HeaderWriterFilter.java:75)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:112)
	at org.springframework.security.web.context.SecurityContextPersistenceFilter.doFilter(SecurityContextPersistenceFilter.java:82)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.context.request.async.WebAsyncManagerIntegrationFilter.doFilterInternal(WebAsyncManagerIntegrationFilter.java:55)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.session.DisableEncodeUrlFilter.doFilterInternal(DisableEncodeUrlFilter.java:42)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
	at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:221)
	at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:186)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:197)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:541)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:135)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:769)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:360)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:399)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:890)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1789)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

**To Reproduce**

- Access a restricted URL
- User is redirected to _login.microsoftonline.com_
- Enter credentials
- User is redirected back to the application
- The exception occurs 

***Code Snippet***

Since version 4.4.0, `JacksonHttpSessionOAuth2AuthorizedClientRepository::getAuthorizedClients` returns an `EmptyMap` instead of an empty `HashMap`. Modifying `authorizedClients` in `JacksonHttpSessionOAuth2AuthorizedClientRepository::saveAuthorizedClient` then fails with an `UnsupportedOperationException`:

```java
Map<String, OAuth2AuthorizedClient> authorizedClients = this.getAuthorizedClients(request); // authorizedClients is an EmptyMap
authorizedClients.put(authorizedClient.getClientRegistration().getRegistrationId(), authorizedClient); // fails with UnsupportedOperationException
````
That is the relevant change in `JacksonHttpSessionOAuth2AuthorizedClientRepository::getAuthorizedClients`:
https://github.com/Azure/azure-sdk-for-java/commit/e9f7f65f47ca17dea3cd161e0b71135cbbebf1b4#diff-87951a4848f4fad213783d16b13938ef38ab11fb6c49475182557d8bab91d04cL79-R80

**Expected behavior**
Login flow should not fail

**Setup (please complete the following information):**
 - OS: Windows
 - IDE: IntelliJ
 - Library/Libraries: com.azure.spring:spring-cloud-azure-starter-active-directory:4.4.0
 - Java version: 11
 - App Server/Environment
  
Configured Oauth 2.0 in `WebSecurityConfigurerAdapter`
```
override fun configure(http: HttpSecurity) {
  http.authorizeRequests()
  ...
  .oauth2Login()
  ...
}
```
Tenant ID, client ID etc. provided in `application.properties`.

 - Frameworks: Spring Boot 2.7.4, Vaadin

**Comment by damir78** [comment]
Hello @mtter ,
this bug was already fixed in #31190  #31213

**Comment by mtter** [comment]
Hi @damir78,

I'm sorry, I missed those. Thanks for the reference 👍
Closing as duplicate of #31213.
